test/test_metadata.py

```diff
@@ -299,3 +299,16 @@ def test_metadata_mapping_update_kw(self):
         self.assertEqual(m['tag2'], 'b')
         m.update(tag2='')
         self.assertIn('tag2', m.deleted_tags)
+
+    def test_metadata_mapping_update_kw_del(self):
+        m = Metadata(tag1='a', tag2='b')
+        del m['tag1']
+
+        m2 = Metadata(tag1='c', tag2='d')
+        del m2['tag2']
+
+        m.update(m2)
```

Comment:
In fact, i'm not sure what we expect here, the patch fixes the fact a tag was marked deleted but still in store, but the result isn't consistent with dict:

dict:

```
>>> m = dict(tag1='a', tag2='b')
>>> del m['tag1']
>>> m2 = dict(tag1='c', tag2='d')
>>> del m2['tag2']
>>> m.update(m2)
>>> print(m)
{'tag2': 'b', 'tag1': 'c'}
```

Metadata:
```
>>> m = Metadata(tag1='a', tag2='b')
>>> del m['tag1']
>>> m2 = Metadata(tag1='c', tag2='d')
>>> del m2['tag2']
>>> m.update(m2)
>>> print(m)
store: {'tag1': ['c']}
deleted: {'tag2'}
images: []
length: 0
```


Here, since `m2` `tag2` was deleted by `del m2['tag2']`, we end with only `tag1` in result: `{'tag1': 'c'}`
I think it's the expected behavior, and all tests pass (including this one), but it would be easier if we actually document `Metadata` expected results & operations, especially when it comes to `deleted_tags`.
BTW, i think this case is buggy in current master.

Reply:
I agree, this behavior makes sense. The old behavior is not very logical. For a operation like copy the expected behavior really should be that you end up with a real copy, that means both store and deleted are the same.

We have cases where we want to preserve the deleted tags for files, but we already handle those in `file.copy_metadata`